### DISCUSSION TITLE
Using LanguageTool for collabora/nextcloud office

### DISCUSSION THREAD

**Discussion by Lextum** [nextcloud/all-in-one#4380] [Ideas] [answered]
Hey,

I have enabled collabora, and it is working fine inside my aio Nextcloud instance (thanks for the awesome work, by the way). I wondered if there was a way to edit the coolwsd.xml inside the collabora container?

My use case would be to enable LanguageTool inside collabora. It could even pair with a [community-containers](https://github.com/nextcloud/all-in-one/tree/main/community-containers) to run the tool locally on the same system.

Thanks again for the outstanding work and sorry if I missed the answered somewhere.

Best regards

**Comment by szaimen** [comment]
Hi, thanks for the positive feedback!

Unfortunately it is currently not possible to edit the coolwsd.xml file in a permanent way.

This Languagetool, is it available as a docker container and if so can you send a link to its docs?

**Reply by mrwsl** [reply]
Would it be possible to implement the coolwsd.xml changes like it was done with the local-ai community image? For the image, a config file is created under the admin user.

**Comment by Lextum** [comment]
I don't think there is an official one, but I am currently using the one by [Erikvl87](https://github.com/Erikvl87/docker-languagetool). The reason for my question about the coolwds.xml is collabora online integration, described here https://sdk.collaboraonline.com/docs/language_tool.html

I am happy to help to integrate it, if you are interested in something like that.

**Reply by szaimen** [reply]
Based on https://github.com/nextcloud/all-in-one/tree/main/community-containers#how-to-add-containers, the containers.json could look as simple as that:
```json
{
    "aio_services_v1": [
        {
            "container_name": "nextcloud-aio-languagetool",
            "display_name": "LanguageTool for Collabora",
            "documentation": "https://github.com/nextcloud/all-in-one/tree/main/community-containers/languagetool",
            "image": "erikvl87/languagetool",
            "image_tag": "latest",
            "internal_port": "8010",
            "restart": "unless-stopped",
            "environment": [
                "TZ=%TIMEZONE%"
            ]
        }
    ]
}
```
However as I said, there is currently no way to configure the collabora container to actually use this. So this would need to be added on top of it...

**Reply by vanilla-extracts** [reply]
Hello, did you find a solution to integrate it within collabora?

**Comment by szaimen** [comment, marked as answer]
Hi, there is a PR for this now: https://github.com/nextcloud/all-in-one/pull/6868

**Reply by szaimen** [reply]
This is now released with v11.9.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by mrwsl** [reply]
Thanks! I tested it on 12.3.0 and I think the extra_params are not passed correctly. Here is a screenshot of my additional settings:
<img width="893" height="374" alt="image" src="https://github.com/user-attachments/assets/c42b72ee-2a62-442d-8115-b557f74cbe1b" />

When I do typos in a document, collabora does not seem to connect to the local languagetool server. I don't see the extra params in the docker container:
```
docker inspect nextcloud-aio-collabora | grep -i extra_params
                "extra_params=--o:ssl.enable=false --o:ssl.termination=true --o:logging.disable_server_audit=true --o:logging.level=warning --o:logging.level_startup=warning --o:welcome.enable=false --o:security.seccomp=true --o:remote_font_config.url=https://tld.com/apps/richdocuments/settings/fonts.json --o:net.post_allow.host[0]=.+",
```
The logs for collabora only show the preloading of languagetool but after that nothing indicates the actual useage of LT.

**Reply by szaimen** [reply]
Can you run `docker inspect nextcloud-aio-collabora | grep -i language` ?

**Reply by mrwsl** [reply]
Sure
```
docker inspect nextcloud-aio-collabora | grep -i language
            "--o:languagetool.enabled=true --o:languagetool.base_url=http://nextcloud-aio-languagetool:8010/v2"
                "--o:languagetool.enabled=true --o:languagetool.base_url=http://nextcloud-aio-languagetool:8010/v2"
```

**Reply by szaimen** [reply]
So yes it gets added but in a different way. 

> but after that nothing indicates the actual useage of LT.

However indeed I am also not sure how to debig this further.

**Reply by mrwsl** [reply]
From a brief look at Collabora's repo, it seems that only 'extra_params' env is used when [launching](https://github.com/CollaboraOnline/online/blob/58c9f41c99d9a0f68339295e946914521e02b1c6/docker/from-source-gh-action/start-collabora-online.sh#L44) coolwsd.

I assume that passing arguments to `start-collabora-online.sh` does nothing:
```
"Path": "/start-collabora-online.sh",
        "Args": [
            "--o:languagetool.enabled=true --o:languagetool.base_url=http://nextcloud-aio-languagetool:8010/v2"
        ],
```
(taken from `docker inspect nextcloud-aio-collabora`.

**Reply by szaimen** [reply]
No, it uses the args via `exec /usr/bin/coolwsd ... "$@"`

**Reply by mrwsl** [reply]
True, the little Copilot button was blocking the view...

**Reply by mrwsl** [reply]
Some further observations:
Like you said, the args are passed correctly to coolwsd
```
docker exec nextcloud-aio-collabora cat /proc/*/cmdline 2>/dev/null | tr '\0' '\n' | grep -A 15 coolwsd
...
/usr/bin/coolwsd
--version
--use-env-vars
--o:sys_template_path=/opt/cool/systemplate
--o:child_root_path=/opt/cool/child-roots
--o:file_server_root_path=/usr/share/coolwsd
--o:cache_files.path=/opt/cool/cache
--o:logging.color=false
--o:stop_on_config_change=true
--o:ssl.enable=false
--o:ssl.termination=true
--o:logging.disable_server_audit=true
--o:logging.level=warning
--o:logging.level_startup=warning
--o:welcome.enable=false
--o:security.seccomp=true
--o:remote_font_config.url=https://tld.com/apps/richdocuments/settings/fonts.json
--o:net.post_allow.host[0]=.+
--o:languagetool.enabled=true --o:languagetool.base_url=https://api.languagetoolplus.com/v2 --o:languagetool.user_name=mail --o:languagetool.api_key=secret
```
Only thing I can see is that the args from the AIO interface are in one single line. When I launch the collabora image locally with the arguments, it seems that they are set correctly
```
docker run --rm -it \
  -p 9980:9980 \
  collabora/code:latest \
  --o:languagetool.enabled=true --o:languagetool.base_url=https://api.languagetoolplus.com/v2 --o:languagetool.user_name=mail --o:languagetool.api_key=secret
  
docker logs <id> 2>&1 | grep -i languagetool
	languagetool.api_key: <redacted>
	languagetool.base_url: https://api.languagetoolplus.com/v2
	languagetool.enabled: true
	languagetool.user_name: <mail>
```

I tested the above with the paid and local languagetool server.

**Reply by szaimen** [reply]
> Only thing I can see is that the args from the AIO interface are in one single line. When I launch the collabora image locally with the arguments, it seems that they are set correctly
> 
> ```
> docker run --rm -it \
>   -p 9980:9980 \
>   collabora/code:latest \
>   --o:languagetool.enabled=true --o:languagetool.base_url=https://api.languagetoolplus.com/v2 --o:languagetool.user_name=mail --o:languagetool.api_key=secret
>   
> docker logs <id> 2>&1 | grep -i languagetool
> 	languagetool.api_key: <redacted>
> 	languagetool.base_url: https://api.languagetoolplus.com/v2
> 	languagetool.enabled: true
> 	languagetool.user_name: <mail>
> ```

This should be improved with https://github.com/nextcloud/all-in-one/pull/7481

**Reply by szaimen** [reply]
> This should be improved with #7481

This is now released with v12.6.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by mrwsl** [reply]
I can confirm that using the paid LanguageTool is now working! Thanks!
